lib/linter/linter.js

```diff
@@ -314,6 +317,64 @@ function createLintingProblem(options) {
     };
 }
 
+/**
+ * Wraps the value in an Array if it isn't already one.
+ * @template T
+ * @param {T|T[]} value Value to be wrapped.
+ * @returns {Array} The value as an array.
+ */
+function asArray(value) {
+    return Array.isArray(value) ? value : [value];
+}
+
+/**
+ * Pushes a problem to inlineConfigProblems if ruleOptions are redundant.
+ * @param {ConfigData} config Provided config.
+ * @param {Object} loc A line/column location
+ * @param {Array} problems Problems that may be added to.
+ * @param {string} ruleId The rule ID.
+ * @param {Array} ruleOptions The rule options, merged with the config's.
+ * @param {Array} ruleOptionsInline The rule options from the comment.
+ * @param {"error"|"warn"} severity The severity to report.
+ * @returns {void}
+ */
+function addProblemIfSameSeverityAndOptions(config, loc, problems, ruleId, ruleOptions, ruleOptionsInline, severity) {
+    const existingConfigRaw = config.rules?.[ruleId];
+
+    if (!existingConfigRaw) {
+        return;
+    }
```

Comment:
One thing I believe we missed discussing in the RFC: should this also report an inline config that turns the rule off when the rule wasn't mentioned in the config? This currently reports an inline config that turns the rule off only if the rule is explicitly disabled (set to `0`/`"off"`) in the config.

For example, an `/* eslint no-undef: 0 */` inline config is reported with this config file:

```js
// eslint.config.js
export default [
    {
        linterOptions: {
            reportUnusedInlineConfigs: "error"
        },
        rules: {
            "no-undef": 0
        }
    }
];


```

but it isn't reported with this config file, although it's effectively unused in the same way:

```js
// eslint.config.js
export default [
    {
        linterOptions: {
            reportUnusedInlineConfigs: "error"
        },
        rules: {
            // there's no "no-undef" rule here
        }
    }
];

```